src/Features/Core/Portable/DisposeAnalysis/DisposableFieldsShouldBeDisposedDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,266 @@
+            {
+                // We only want to analyze types which are disposable (implement System.IDisposable directly or indirectly)
+                // and have at least one disposable field.
+                var namedType = (INamedTypeSymbol)symbolStartContext.Symbol;
+                if (!namedType.IsDisposable(disposeAnalysisHelper.IDisposable))
+                {
+                    return;
+                }
+
+                var disposableFields = disposeAnalysisHelper.GetDisposableFields(namedType);
+                if (disposableFields.IsEmpty)
+                {
+                    return;
+                }
+
+                var analyzer = new SymbolAnalyzer(disposableFields, disposeAnalysisHelper);
+
+                // Register an operation block action to analyze disposable assignments and dispose invocations for fields.
+                symbolStartContext.RegisterOperationBlockStartAction(analyzer.OnOperationBlockStart);
+
+                // Register symbol end action for containing type to report non-disposed fields.
+                // We report fields that have disposable type (implement System.IDisposable directly or indirectly)
+                // and were assigned a disposable object within this containing type, but were not disposed in
+                // containing type's Dispose method.
+                symbolStartContext.RegisterSymbolEndAction(analyzer.OnSymbolEnd);
+            }
+
+            private void AddOrUpdateFieldDisposedValue(IFieldSymbol field, bool disposed)
+            {
+                Debug.Assert(_disposableFields.Contains(field));
+                Debug.Assert(!field.IsStatic);
+                Debug.Assert(field.Type.IsDisposable(_disposeAnalysisHelper.IDisposable));
+
+                // Update the dispose value for the field.
+                // Update value factory delegate ensures that fields for which we have
+                // already seen dispose invocations, i.e. currentValue = true, continue to be marked as disposed.
+                _fieldDisposeValueMap.AddOrUpdate(field,
+                    addValue: disposed,
+                    updateValueFactory: (f, currentValue) => currentValue || disposed);
+            }
+
+            private void OnSymbolEnd(SymbolAnalysisContext symbolEndContext)
+            {
+                foreach (var kvp in _fieldDisposeValueMap)
+                {
+                    IFieldSymbol field = kvp.Key;
+                    bool disposed = kvp.Value;
+                    if (!disposed)
+                    {
+                        // Disposable field '{0}' is never disposed
+                        var diagnostic = Diagnostic.Create(s_disposableFieldsShouldBeDisposedRule, field.Locations[0], field.Name);
+                        symbolEndContext.ReportDiagnostic(diagnostic);
+                    }
+                }
+            }
```

Comment:
do you remember whether this symbol analyzer now get deleted after this point? or all symbol analyzers alive until whole compilationwithAnalyzer get deleted?

Reply:
Nested actions registered within SymbolStart callback can never be called after SymbolStart. So yes, the driver will not maintain any state pertaining to Symbol associated with SymbolStart/End once its SymbolEnd action has been called.